plugins/logitech-rallysystem/fu-logitech-rallysystem-tablehub-device.c

```diff
@@ -0,0 +1,426 @@
+
+struct _FuLogitechRallySystemTableHubDevice {
+	FuUsbDevice parent_instance;
+	guint bulk_ep[EP_LAST];
+	guint bulk_iface;
+};
+
+/* tablehub firmware image state */
+enum image_state { img_new, img_valid, img_invalid };
+
+/* tablehub firmware version info */
+struct fw_img_info {
+	char fw_version[VERSION_LENGTH];
+	enum image_state img_state;
+};
+
+/* bulk command */
+struct logi_cmd {
+	guint16 id;
+	guint16 status;
+	guint32 len;
+};
+#define LOGI_CMD_SIZE (sizeof(struct logi_cmd))
+
+struct progress_data {
+	guint32 completed;
+};
+
+struct download_packet {
+	struct logi_cmd cmd;
+	char fw_version[VERSION_LENGTH];
+};
+
+G_DEFINE_TYPE(FuLogitechRallySystemTableHubDevice,
+	      fu_logitech_rallysystem_tablehub_device,
+	      FU_TYPE_USB_DEVICE)
+
+static void
+fu_logitech_rallysystem_tablehub_device_to_string(FuDevice *device, guint idt, GString *str)
+{
+	FuLogitechRallySystemTableHubDevice *self = FU_LOGITECH_RALLYSYSTEM_TABLEHUB_DEVICE(device);
+	fu_string_append_kx(str, idt, "IfaceId", self->bulk_iface);
+	fu_string_append_kx(str, idt, "EpBulkIn", self->bulk_ep[EP_IN]);
+	fu_string_append_kx(str, idt, "EpBulkOut", self->bulk_ep[EP_OUT]);
+}
+
+static gboolean
+fu_logitech_rallysystem_tablehub_device_probe(FuDevice *device, GError **error)
+{
+	FuLogitechRallySystemTableHubDevice *self = FU_LOGITECH_RALLYSYSTEM_TABLEHUB_DEVICE(device);
+	g_autoptr(GPtrArray) intfs = NULL;
+
+	intfs = g_usb_device_get_interfaces(fu_usb_device_get_dev(FU_USB_DEVICE(self)), error);
+	if (intfs == NULL)
+		return FALSE;
+
+	for (guint i = 0; i < intfs->len; i++) {
+		GUsbInterface *intf = g_ptr_array_index(intfs, i);
+		if (g_usb_interface_get_class(intf) == G_USB_DEVICE_CLASS_VENDOR_SPECIFIC) {
+			g_autoptr(GPtrArray) endpoints = g_usb_interface_get_endpoints(intf);
+			self->bulk_iface = g_usb_interface_get_number(intf);
+			if (endpoints == NULL)
+				continue;
+			for (guint j = 0; j < endpoints->len; j++) {
+				GUsbEndpoint *ep = g_ptr_array_index(endpoints, j);
+				if (j == EP_OUT)
+					self->bulk_ep[EP_OUT] = g_usb_endpoint_get_address(ep);
+				else
+					self->bulk_ep[EP_IN] = g_usb_endpoint_get_address(ep);
+			}
+		}
+	}
+	fu_usb_device_add_interface(FU_USB_DEVICE(self), self->bulk_iface);
+	g_debug("iface: %u ep_out: %u ep_in: %u",
+		self->bulk_iface,
+		self->bulk_ep[EP_OUT],
+		self->bulk_ep[EP_IN]);
+	return TRUE;
+}
+
+static gboolean
+fu_logitech_rallysystem_tablehub_device_send(FuLogitechRallySystemTableHubDevice *self,
+					     guint8 *req_buf,
+					     guint32 req_buf_len,
+					     GError **error)
+{
+	gsize transferred = 0;
+	GCancellable *cancellable = NULL;
+	g_return_val_if_fail(req_buf != NULL, FALSE);
+	if (!g_usb_device_bulk_transfer(fu_usb_device_get_dev(FU_USB_DEVICE(self)),
+					self->bulk_ep[EP_OUT],
+					req_buf,
+					req_buf_len,
+					&transferred,
+					FU_LOGITECH_RALLYSYSTEM_TABLEHUB_DEVICE_IOCTL_TIMEOUT,
+					cancellable,
+					error) ||
+	    (req_buf_len != transferred)) {
```

Comment:
i don't think this works -- the `req_buf_len != transferred` needs to be a new conditional.